### PR TITLE
Add tutorial examples of per-channel quantization

### PR DESCRIPTION
**Pull request by OscarSavolainenDR**
## Purpose of this PR
To expand QuantAct notebook with scaling per output channel examples, as referenced in issue #862 

## Changes made in this PR
We add the example of a `QuantReLU`, and show how per-channel quantization can enable one to better match the dynamic range of any individual channel.

**Note:** I am not sure my explanation of  `per_channel_broadcastable_shape`  and `scaling_stats_permute_dims`  are correct:
```
`per_channel_broadcastable_shape` represents what the dimensions of the quantization parameters will be, 
and should be  laid out to match those of the output channels of the outputted tensor. We also need to 
specify the permutation dimensions via `scaling_stats_permute_dims` so as to shape the tensor into a
standard format of output channels first. This is so that during the statistics gathering stage of QAT the
correct stats will be gathered.
```

### CODE REVIEW

**Comment by OscarSavolainenDR** [comment]
@Giuseppe5 I have added some details, let me know if I should add some more! A lot of the permutation code relies on boilerplate stuff, I wasn't sure how detailed of a tracing of the boilerplate I should include in the tutorial. I have all my notes so can get quite granular if we want to.

File: notebooks/02_quant_activation_overview.ipynb [hunk 1]
```diff
@@ -664,25 +664,233 @@
...
+    "per_tensor_quant_relu = QuantReLU(return_quant_tensor=True)\n",
+    "out_tensor = per_tensor_quant_relu(inp3)\n",
+    "out_tensor.scale * ((2**8) -1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can see that the per-tensor scale parameter has calibrated itself to provide a full quantization range of 3, matching that of the most extreme channel. \n",
```
Comment (Giuseppe5) [inline notebooks/02_quant_activation_overview.ipynb:714]: Most extreme -> highest? If I'm understanding the meaning correctly

File: notebooks/02_quant_activation_overview.ipynb [hunk 2]
```diff
@@ -664,25 +664,233 @@
...
+    "\n",
+    "We can take a look at the `QuantReLU` object, and in particular look at what the ` scaling_impl` object is composed of. It is responsible for gathering statistics for determining the quantization parameters, and we can see that its ` stats_input_view_shape_impl` attribute is set to be an instance of `OverTensorView`. This is defined [here](https://github.com/Xilinx/brevitas/blob/200456825f3b4b8db414f2b25b64311f82d3991a/src/brevitas/core/function_wrapper/shape.py#L78), and serves to flatten out the observed tensor into a 1D tensor and, in this case, use the `AbsPercentile` observer to calculate the quantization parameters during the gathering statistics stage of QAT."
```
Comment (Giuseppe5) [inline notebooks/02_quant_activation_overview.ipynb:716]: There is one extra space:
` scaling_impl` -> `scaling_impl`, ` stats_input_view_shape_impl`->`stats_input_view_shape_impl`

File: notebooks/02_quant_activation_overview.ipynb [hunk 3]
```diff
@@ -664,25 +664,233 @@
...
+   "metadata": {},
+   "source": [
+    "Next, we initialise a new `QuantRelU` instance, but this time we specify that we desire per-channel quantization i.e. `scaling_per_output_channel=True`. This will implictly call `scaling_stats_input_view_shape_impl`, defined [here](https://github.com/Xilinx/brevitas/blob/200456825f3b4b8db414f2b25b64311f82d3991a/src/brevitas/quant/solver/common.py#L184), and will change the `QuantReLU` from using a per-tensor view when gathering stats to a per output channel view ([`OverOutputChannelView`](https://github.com/Xilinx/brevitas/blob/200456825f3b4b8db414f2b25b64311f82d3991a/src/brevitas/core/function_wrapper/shape.py#L52)). This simply permutes the tensor into a 2D tensor, with dim 0 equal to the number of output channels.\n",
+    "\n",
+    "To accomplish this, we also need to give it some extra information: `scaling_stats_permute_dims` and `per_channel_broadcastable_shape`. `scaling_stats_permute_dims` is responsible for defining how we do the permutation. `per_channel_broadcastable_shape` simply represents what the dimensions of the quantization parameters will be, i.e. there should be one parameter per output channel.\n",
```
Comment (Giuseppe5) [inline notebooks/02_quant_activation_overview.ipynb:785]: `per_channel_broadcastable_shape` is necessary to understand which along which dimensions the scale factor has to be broadcasted to match the dimensions of the input.
By default, pytorch will broadcast along the first rightmost dimension for which the shapes of the two tensors matches. To make sure that we apply the scale factor in our desired dimension, we need to tell PyTorch how to correctly broadcast the scale factors.

**Comment by Giuseppe5** [comment]
Could you please rebase this over dev? Thanks!

File: notebooks/02_quant_activation_overview.ipynb (hunk 1 above)
Comment (OscarSavolainenDR) [reply notebooks/02_quant_activation_overview.ipynb:714]: Yep, I'll make it more clear!

File: notebooks/02_quant_activation_overview.ipynb (hunk 3 above)
Comment (OscarSavolainenDR) [reply notebooks/02_quant_activation_overview.ipynb:785]: Awesome, will add more context!

**Comment by OscarSavolainenDR** [comment]
I'm unusually busy until Wednesday, but will rebase and adjust the PR along the lines of the comments then!

File: notebooks/02_quant_activation_overview.ipynb (hunk 3 above)
Comment (OscarSavolainenDR) [reply notebooks/02_quant_activation_overview.ipynb]: I took your description, as it was a very good explanation haha, and added it to the PR!

File: notebooks/02_quant_activation_overview.ipynb (hunk 2 above)
Comment (OscarSavolainenDR) [reply notebooks/02_quant_activation_overview.ipynb]: Done.

File: notebooks/02_quant_activation_overview.ipynb (hunk 1 above)
Comment (OscarSavolainenDR) [reply notebooks/02_quant_activation_overview.ipynb]: I changed it to indicate the channel with largest dynamic range, instead of most "extreme".

**Pull request merged**